File: 019_Length of Last Word.cpp
/*
Length of Last Word

Given a string s consisting of words and spaces, return the length of the last word in the string.

Strategy: Trim trailing spaces and find last token length in one pass from end.
*/

#include <string>
#include <iostream>
using namespace std;

int lengthOfLastWord(const string& s) {
    int i = s.size() - 1;
    while (i >= 0 && s[i] == ' ') --i;
    int len = 0;
    while (i >= 0 && s[i] != ' ') { ++len; --i; }
    return len;
}

int main() {
    cout << "LenLast 1: " << lengthOfLastWord("Hello World") << '\n';
    cout << "LenLast 2: " << lengthOfLastWord("   fly me   to   the moon  ") << '\n';
    cout << "LenLast 3: " << lengthOfLastWord("luffy is still joyboy") << '\n';
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 019_Length of Last Word.py
"""
Length of Last Word

Return the length of the last word in a string.
"""

def length_of_last_word(s: str) -> int:
    i = len(s) - 1
    while i >= 0 and s[i] == ' ': i -= 1
    length = 0
    while i >= 0 and s[i] != ' ':
        length += 1
        i -= 1
    return length


if __name__ == '__main__':
    print('LenLast 1:', length_of_last_word('Hello World'))
    print('LenLast 2:', length_of_last_word('   fly me   to   the moon  '))
    print('LenLast 3:', length_of_last_word('luffy is still joyboy'))

# Time Complexity: O(n)
# Space Complexity: O(1)
